### ISSUE TITLE
JavaKerberos authentication on Spark framework

### ISSUE THREAD

**Issue by quarkonium3** [Microsoft/mssql-jdbc#820]
## Driver version
7.0.0 also 7.1.1 preview

## SQL Server version
SQL 2016 also 2014 and 2012

## Client Operating System
Ubuntu 18.04.1 on each node of a Spark cluster

## JAVA/JVM version
Java 1.8

## Table schema
N/A

## Problem description
I am trying to get a Spark cluster to write to SQL server using JavaKerberos (i.e., I specify `integratedSecurity=true;authenticationScheme=JavaKerberos` in the connection string) with credentials specified in a keyTab file. I am not having much success (the problem is the same if I specify credentials in the connections string). 

I am submitting the job to spark (a YARN mode, 4-node, in-house spark version 2.3.0 cluster) with:
``` 
spark-submit --driver-class-path mssql-jdbc-7.0.0.jre8.jar \
--jars /path/to/mssql-jdbc-7.0.0.jre8.jar \
--conf spark.executor.extraClassPath=/path/to/mssql-jdbc-7.0.0.jre8.jar \
--conf "spark.driver.extraJavaOptions=-Djava.security.auth.login.config=/path/to/SQLJDBCDriver.conf" \
--conf "spark.executor.extraJavaOptions=-Djava.security.auth.login.config=/path/to/SQLJDBCDriver.conf" \
application.jar
```
where application.jar is a plain scala spark application. A colleague has replicated the same problem in python with PySpark however.
 
The application ‘half’ works in that the spark driver program will authenticate correctly and create a table, however any of the executors fail to write any data to the table with an exception:

`
java.security.PrivilegedActionException: GSSException: No valid credentials provided (Mechanism level: Failed to find any Kerberos tgt)
`

Things I found:
* I can get everything to work if I specify SQL server credentials (however I need to use integrated security in my application)
* The keytab and login module file “SQLJDBCDriver.conf” seem to be specified correctly since they work for the driver
* I can see the executors pick up the correct command line options in the spark UI : 
-Djava.security.auth.login.config=/path/to/SQLJDBCDriver.conf
* Using some of the logging options it looks like the executors read the SQLJDBCDriver.conf file but then ignore it.

After a lot of debugging I notice the different behaviour between driver and executors comes from the following lines in KerbAuthentication.java
https://github.com/Microsoft/mssql-jdbc/blob/709342fbedb2e35f25c02c7f3104658319ff5472/src/main/java/com/microsoft/sqlserver/jdbc/KerbAuthentication.java#L83-L90

I noticed:
* It works for the driver because it receives null for currentSubject and then creates a new LoginContext using the credentials from the keytab
* On the other hand, for some reason, the executors at this point receive a currentSubject populated with the local user credentials of the spark cluster and _not_ those specified in the keytab (or in the connection string if I use that method)

I don't really understand why this is happening – it seems as if the executors are picking up some context object before they know they need to create one with different credentials. I can’t seem to get it to use the keytab credentials after trying many, many variations.

As an experiment I recompiled the driver,  nulling out currentSubject on line 84. This then works – a new LoginContext is created but this time with credentials as specified in the keytab and all data is written to the table correctly.

Can anyone help me understand how I can force the executors to use credentials provided in the keytab (or connection string) or otherwise get JavaKerberos to work with Spark?

**Comment by quarkonium3** [comment]
Hi @peterbae, many thanks for looking into this. Likewise, I am not sure it is necessarily a jdbc driver problem and not something spark is doing. Or even that it's something more difficult like perhaps it's the way spark executors have to start that the jdbc driver doesn't like.

The jdbc driver behaviour is still a bit puzzling however. With the java command line options I have selected I _have_ to have a SQLJDBCDriver.conf file available or it fails with file not found. The file _has_ to be correctly formatted or it fails with 'no login modules available for SQLJDBCDriver.conf'. With debug logging options switched on I can even see that this file is being read and the wanted options are parsed. _However_ it then completely disregards what it's read and tries to use default properties and the local user credentials.
  
In fact the  (GSSException: No valid credentials provided (Mechanism level: Failed to find any Kerberos tgt)) does make some kind of sense given that it is trying to use the wrong credentials. I can get this message if I deliberately use wrong credentials and use the default driver behaviour i.e., `useTicketCache=true useKeyTab=false` . 

I am pretty confident our Kerberos setup is not misconfigured - I can test TGT generation and so forth with kinit, klist etc. And as I said with my nasty hack it does work and I can observe the correct logins on the SQL server.

I guess if you have any idea on where the context it gets from `AccessControlContext context = AccessController.getContext();` on line 83 and the embedded subject it gets from line 84 is originally being set it would be helpful.

I do plan to take this up somewhere in the spark ecosystem as well to see if I can gain more insight - just haven't got there yet. This is eating up a vast amount of time, but is a blocking issue for this application so will soldier on...

I will try and build a standalone java application as you suggest to see if I can get some more understanding of the issue. I fear it will work in the same way the spark driver program works, rather than the problematic executors (sorry about the confusing terminology there: we have a spark driver and jdbc driver). 

Anyway - thanks again

**Comment by quarkonium3** [comment]
Well - apologies for the false alarm. It seems spark executors cannot do Kerberos auth according to an old JIRA (https://issues.apache.org/jira/browse/SPARK-12312) which is still true of latest spark v2.3.2 according to the spark user list and some more testing I did.

Seems like a pretty serious limitation to me, not sure why it's never been addressed. Anyway, I've worked around it for now by creating a wrapper for the jdbc driver which manages the keytab auth and passes back the connection to spark based on the idea here: https://datamountaineer.com/2016/01/15/spark-jdbc-sql-server-kerberos/
